### DISCUSSION TITLE
Box_intersection_d:  Can we make the Callback a const reference ?

### DISCUSSION THREAD

**Discussion by afabri** [CGAL/cgal#6456] [Ideas]
## Issue Details

In the package for fast box intersections in the [API ](https://doc.cgal.org/latest/Box_intersection_d/group__PkgBoxIntersectionD__box__intersection__d.html#ga7ac96673d25a437c98e8abb864a3362d)we write  `Callback callback`.       Is it acceptable to turn that into `const Callback& callback`?   I would like to do that for performance reasons.   I would also document it.   This may imply changes on the users code. Even in the PMP package I have to make some data members of callbacks mutable, as the `Callback::operator()(Box*,Box*)` must be `const`.   Alternatvely, we could switch to `Callback& callback`.

**Comment by lrineau** [comment]
Using `Callback& callback` will break code where the callback is not a l-value.

Using `const Callback&` will force the callback to be const, and will break user-code, including the examples in `Box_intersection_d/examples`. If there must be a redirection to a big object, then let it be in the callback object: callbacks should be lightweight objects, like property maps.

**Comment by mglisse** [comment]
`Callback&& callback`?

**Comment by sloriot** [comment]
actually, passing `std::ref(callback)` directly is working because there is a forward to the wrapped type of all `operator()` in `std::reference_wrapper`. However I did not notice any difference for corefinement.
